### PR TITLE
Prioritizes fast switch to master in HA

### PR DESCRIPTION
**Pull request by tinwelint**
Previously as part of switching to master there was waiting introduced to
let active transactions complete normally before switching. The concept of
doing that is nice, but that comes at an expense of cluster downtime when
there will be times where the slave that was picked to become master could
wait for a long time for active transactions that perhaps would never
complete or was blocked on some condition related to the previous master
no longer being available. Reasons why having master switching wait is
bad:
- A master in a cluster is very important, without it the cluster cannot
  process any requests
- Awaiting open transactions to complete assumes that this instance just
  now was a slave that is switching to master, which means the previous
  master where these active transactions were hosted is no longer available
  so these open transactions cannot continue and complete anyway,
  so what's the point waiting for them?
- Read transactions may still be able to complete, but the correct
  response to failures in those is to have them throw transient error exceptions
  hinting that they should be retried, at which point they may get redirected
  to another instance, or to this instance if it has completed the switch until then.

There was a related wait in SlaveLocksClient which was there to bridge
some gap if slave in the middle of a transaction found itself unavailable.
This waiting would wait for the slave to become available again, with the
hope that the same cluster member was still master and hadn't restarted
from last time this transaction saw the master. A better way to handle
this is to make sure such transaction threads gets notified about this
fact via a transient error exception, so that the transaction can be
retried. Another problem with waiting at this point in time was that
already acquired locks would be helt during a long time, blocking other
transactions in the cluster.

While doing this the setting `ha.state_switch_timeout` was split into that
and `ha.internal_state_switch_timeout`. Previously that timeout setting
was used both for timeout for request threads as well as timeout for
internal switching logic, which was just a mixup that shouldn't have been
to begin with. This means that members that switch to slave waits for
active transactions to complete, using `ha.internal_state_switch_timeout`
and requests wanting to begin transaction or similar waits for db to be
available using `ha.state_switch_timeout`, which is much longer.

### CODE REVIEW

**Comment by thobe** [comment]
This is quite a substantial change in behaviour. I'd like to see thorough testing of what the implications of these changes are to users of the surface APIs, not just to the behaviour/stability of the cluster.

**Comment by tinwelint** [comment]
Let's conduct such experiments

File: enterprise/ha/src/main/java/org/neo4j/kernel/ha/cluster/SwitchToMaster.java [hunk 1]
```diff
@@ -95,44 +89,38 @@ public URI switchToMaster( LifeSupport haCommunicationLife, URI me )
...
+        //   at which point they may get redirected to another instance, or to this instance if it has completed
+        //   the switch until then.
 
-        /*
-         * Synchronizing on the xaDataSourceManager makes sense if you also look at HaKernelPanicHandler. In
-         * particular, it is possible to get a masterIsElected while recovering the database. That is generally
-         * going to break things. Synchronizing on the xaDSM as HaKPH does solves this.
-         */
-        //noinspection SynchronizationOnLocalVariableOrMethodParameter
-//        synchronized ( xaDataSourceManager )
```
Comment (thobe) [inline enterprise/ha/src/main/java/org/neo4j/kernel/ha/cluster/SwitchToMaster.java:111]: silly old code...

File: enterprise/ha/src/main/java/org/neo4j/kernel/ha/HaSettings.java [hunk 2]
```diff
@@ -47,9 +47,14 @@
...
     public static final Setting<Long> read_timeout = setting( "ha.read_timeout", DURATION, "20s" );
 
-    @Description( "Timeout for waiting for instance to become master or slave." )
+    @Description( "Timeout for request threads waiting for instance to become master or slave." )
     public static final Setting<Long> state_switch_timeout = setting( "ha.state_switch_timeout", DURATION, "120s" );
 
+    @Description( "Timeout for waiting for internal conditions during state switch, like for transactions "
+            + "to complete, before switching to master or slave." )
+    public static final Setting<Long> internal_state_switch_timeout =
+    setting( "ha.internal_state_switch_timeout", DURATION, "10s" );
```
Comment (chrisvest) [inline enterprise/ha/src/main/java/org/neo4j/kernel/ha/HaSettings.java]: Odd formatting. I'd line-break after `setting(` and indent the parameters. Is 10 seconds not quite long, with a default heartbeat timeout of 11 seconds?

Comment (tinwelint) [inline enterprise/ha/src/main/java/org/neo4j/kernel/ha/HaSettings.java]: Somehow the indenting was lost. Also all the other line breaking in that file is after the `=`

This timeout is unrelated to heartbeats though. This is the max time a slave will wait between getting elected and starting the switch to master, waiting for active transaction to complete. The longer this timeout, the longer we've got a master-less cluster.

File: enterprise/ha/src/main/java/org/neo4j/kernel/ha/cluster/SwitchToSlave.java [hunk 3]
```diff
@@ -262,8 +263,9 @@ public URI switchToSlave( LifeSupport haCommunicationLife, URI me, URI masterUri
 
         monitor.switchToSlaveStarted();
 
-        // Wait for current transactions to stop first
-        long deadline = SYSTEM_CLOCK.currentTimeMillis() + config.get( HaSettings.state_switch_timeout );
+        // Wait a short while for current transactions to stop first, just to be nice.
+        // We can't wait forever since switching to our designated role is quite important.
+        long deadline = SYSTEM_CLOCK.currentTimeMillis() + config.get( HaSettings.internal_state_switch_timeout );
```
Comment (thobe) [inline enterprise/ha/src/main/java/org/neo4j/kernel/ha/cluster/SwitchToSlave.java:268]: I guess a fair question would be: why wait at all? What is it that these transactions will be able to do in that time that we are waiting for them?

It is pretty fair to assume that the master is gone here, so they cannot write. So the only transactions that _can_ finish are read-only transactions. Am I correct?

I'm wondering if it could be better to mark the transactions with some sort of failure mark that permits read operations for transactions that have not performed any write operations before (to allow read-only transactions to continue), but fails any write operations, or any operations at all on transactions that have already performed write operations.

With such a change we could perhaps completely remove the wait.

Granted, that change would be quite a lot bigger...

**Comment by thobe** [comment]
Am I understanding the situation correctly that a master-less cluster is essentially just a bunch of zombie machines unable to do much at all?

**Comment by tinwelint** [comment]
Yes, the slaves will have their availability guard raised so even read transactions cannot really begin.

File: enterprise/ha/src/main/java/org/neo4j/kernel/ha/cluster/SwitchToSlave.java (hunk 3 above)
Comment (tinwelint) [inline enterprise/ha/src/main/java/org/neo4j/kernel/ha/cluster/SwitchToSlave.java:268]: True, that sounds like a reasonable thing to do. 10 seconds is an just an arbitrary amount of seconds, although certainly less than 120 which the case before this change. And since we've concluded that it's only read transactions that we are waiting for then 10 seconds is a pretty long time.

Comment (tinwelint) [inline enterprise/ha/src/main/java/org/neo4j/kernel/ha/cluster/SwitchToSlave.java:268]: We should perhaps get this in as-is, to keep changesets small to be able to more easily see incremental changes in builds further on.

**Comment by davidegrohmann** [comment]
@tinwelint what are the experimental results for the change of behavior?

**Comment by tinwelint** [comment]
It's hard to compare since ha robustness testing without this fix fails so early. Although failure distribution, i.e. which type of transaction failures occur and how often, look roughly the same both before and after. To me it looks like this only changes things for the better. Perhaps you could do some additional comparisons as well?
